### DISCUSSION TITLE
How to sharing viewmodel between an activity and each navgraph destination

### DISCUSSION THREAD

**Discussion by mikucat0309** [InsertKoinIO/koin#1436] [Q&A] [answered]
I already read [this](https://insert-koin.io/docs/reference/koin-android/viewmodel#navigation-graph-viewmodel-313) document, but it only talks about sharing between destination, in other word, how to get same viewmodel under same acitivity, no matter whether in composable function or navgraph destination.

BTW, `koinNavGraphViewModel` can not work with Jetpack Compose, since it doesn't need fragments.

**Comment by mikucat0309** [comment, marked as answer]
```kotlin
@Composable
inline fun <reified T : ViewModel> getActivityViewModel(): T =
    getViewModel(owner = LocalContext.current as ComponentActivity)
```

**Comment by PGautam27** [comment]
One of the best alternative to tackle this problem is to use dagger hilt. Which is splendid, as hiltViewModel would take care of this problem. It doesn't matter whether it's a navgraph or in some composables. I'll add the dependencies and a sample code below.

Add this in the plugin in gradle
```
plugins {
    ....
    id 'kotlin-kapt'
    id 'dagger.hilt.android.plugin'
}
```

Add this as a dependency
```
dependencies {
    //Dagger - Hilt
    implementation "com.google.dagger:hilt-android:2.38.1"
    kapt "com.google.dagger:hilt-android-compiler:2.37"
    implementation "androidx.hilt:hilt-lifecycle-viewmodel:1.0.0-alpha03"
    kapt "androidx.hilt:hilt-compiler:1.0.0"
    implementation 'androidx.hilt:hilt-navigation-compose:1.0.0-alpha03'` 
}
```

Add this to your classpath
```
dependencies{
       classpath "com.google.dagger:hilt-android-gradle-plugin:2.38.1"
}
```

Once the you've successfully added this to your gradle files. Then in your composables use these
```
@Composable
fun compScreen(
    viewModel: ComplViewModel = hiltViewModel()
) {
   //your functions
}
```

now the viewModel is directly injected to your composable. I hope this helps you.

Happy coding!!

**Reply by mikucat0309** [reply]
@PGautam27 Your answer should not work, NavHost provides last NavBackStackEntry as ViewModelStoreOwner.

https://github.com/androidx/androidx/blob/2bbe1a293edea26f8af9a8c9dc490e70207d62af/navigation/navigation-compose/src/main/java/androidx/navigation/compose/NavHost.kt#L162-L164

That is, ViewModels in destination bind to each destination, not NavGraph, NavHost or Activity, ViewModels get in destination will not equal to ones get in Activity, sharing ViewModel between destination also not work.

**Reply by PGautam27** [reply]
I totally agree to this fact. Here since we are using hiltviewnodel(). We are injecting the viewmodle and not sharing them.

**Reply by PGautam27** [reply]
You could check upto Philip lackner videos from youtube if you are free. I already said. It's just an alternative of doing. And not the exact solution.
